server/src/main/java/io/crate/metadata/ColumnIdent.java

```diff
@@ -196,6 +197,24 @@ public boolean isChildOf(ColumnIdent parent) {
         return false;
     }
 
+    /**
+     * Replaces 'oldName' starting from left most side of the column to 'newName'.
+     * e.g.:: Invoking rename("a.b", "a.x") on "a.b.c.d" returns "a.x.c.d"
+     *                 rename("a", "b")     on "a"       returns "b"
+     */
+    @NotNull
+    public ColumnIdent rename(@NotNull ColumnIdent oldName, @NotNull ColumnIdent newName) {
+        if (this.equals(oldName)) {
+            return newName;
+        }
+        if (this.isChildOf(oldName)) {
+            var newPath = new ArrayList<>(newName.path());
+            newPath.addAll(this.path().subList(oldName.path().size(), this.path().size()));
+            return new ColumnIdent(newName.name(), newPath);
+        }
+        return this;
+    }
+
```

Comment:
I find the design of this strange. Renaming a `ColumnIdent` instance from an old name to a new name doesn't really make sense to me. The `newName` parameter is already the new name?

The function actually has a "rename-if-(sub)match" behavior but I don't understand why the path sub-replacement is needed.

It also seems a bit of a waste that we re-create _all_ `Reference` instances via their rename, despite only expecting to change the name of _one_ column.


Wouldn't it be more straightforward to re-use the `withReferenceIdent` - which already gives a Reference a new name - and use a `ref.column().equals(oldColumn) ? ref.withReferenceIdent(...) : ref` in the RenameTableTask?

Reply:
> I find the design of this strange. Renaming a ColumnIdent instance from an old name to a new name doesn't really make sense to me. The newName parameter is already the new name?

If an object is renamed, the sub-columns need to be renamed as well so newName may need to be appended with child names.

> Wouldn't it be more straightforward to re-use the withReferenceIdent - which already gives a Reference a new name - and use a ref.column().equals(oldColumn) ? ref.withReferenceIdent(...) : ref in the RenameTableTask?

`Reference::rename` would propagate rename calls to its `Reference` fields which is more than what `withReferenceIdent` does. Would you like the the two methods to be merged? Also, could you explain why it should happen in `RenameColumnTask` instead of `DocTableInfo` ?

Reply:
> If an object is renamed, the sub-columns need to be renamed as well so newName may need to be appended with child names.

Couldn't we make this a bit more concrete and explicit then?

> Reference::rename would propagate rename calls to its Reference fields which is more than what withReferenceIdent does

It seems to me as this is mixing a few different aspects all in the same - somewhat misleading - function.

- `Reference` can have a function which changes the name/ColumnIdent of it. This should of course update the name in _all_ of its relevant properties. If `withReferenceIdent` isn't doing that, that could be a bug

- _Other_ references might have had references to the renamed reference. They need an update too, but that's different than renaming a Reference

- The "if-match" aspect.





> Also, could you explain why it should happen in RenameColumnTask instead of DocTableInfo ?

I just mentioned `RenameColumnTask` because that's the entry point. Doing it in `DocTableInfo` is fine.

---

I kinda get that it fit's the current use-case, but the design still feels rather unusual. E.g. if you focus on the signature: `rename(ColumnIdent from, ColumnIdent newName) -> ColumnIdent` as instance method on a `ColumnIdent` - it's very hard to understand and confusing, because `newName` isn't the actual `newName` (for the object case you mentioned)

Reply:
I'd also be interested what @seut or @matriv think about this. If it only bothers me, maybe it's fine to leave as is.

Reply:
I agree with @mfussenegger about the mixed scope. I'd also prefer to separate them at e.g. `DocTableInfo.renameColumn` by something like:
```

Map<Reference, Reference> renamedReferences = references.values().stream()
  .filter(ref -> ident-is-equal-or-child-of-old-name)
  .collect(Collectors.toMap(ref, Reference.withReferenceIdent(ref.ident())));
```

And afterwards replace dependencies by using
```
 Function<Symbol, Symbol> updateRef = symbol -> RefReplacer.replaceRefs(symbol, ref -> renamedReferences.getOrDefault(ref, ref));
```

Reply:
Thank you all for your inputs, please check 2d076a9d5f644a40adf1798f2364d6ef1d60fe49 for update.